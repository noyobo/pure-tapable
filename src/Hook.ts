import { DoneCallback } from './types.js';

abstract class Hook<A extends unknown[]> {
  protected readonly args: string[];

  protected constructor(args: string[] = []) {
    this.args = args;
    this.validateArgs?.();
  }

  /**
   * 获取运行参数，受实例化参数决定
   * @private
   */
  protected argsToParams(params: A): A {
    return this.args.map((name, index) => {
      return params[index];
    }) as A;
  }

  protected validateArgs?(): void;
}

/**
 * 同步 Hook 基类
 */
export abstract class Sync<A extends unknown[], F> extends Hook<A> {
  protected readonly taps: Array<{ name: string; fn: F }>;

  /**
   * 钩子参数
   * @param args
   */
  constructor(args?: string[]) {
    super(args);
    this.taps = [];
  }

  tap(name: string, fn: F): void {
    this.taps.push({ name, fn });
  }

  // 由继承者实现
  abstract call(...args: unknown[]): any;
}

/**
 * 异步 Hook 基类
 */
export abstract class Async<A extends unknown[], F, P> extends Hook<A> {
  protected readonly asyncTaps: Array<{ name: string; fn: F }>;
  protected readonly promiseTaps: Array<{ name: string; fn: P }>;

  constructor(args?: string[]) {
    super(args);
    this.asyncTaps = [];
    this.promiseTaps = [];
  }

  tapAsync(name: string, fn: F) {
    this.asyncTaps.push({ name, fn });
  }

  tapPromise(name: string, fn: P) {
    this.promiseTaps.push({ name, fn });
  }

  abstract callAsync(...args: [...A, DoneCallback]): any;

  abstract promise(...args: A): Promise<any>;
}
