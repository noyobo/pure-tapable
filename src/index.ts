export { default as SyncHook } from './SyncHook';
export { default as SyncBailHook } from './SyncBailHook';
export { default as SyncWaterfallHook } from './SyncWaterfallHook';
export { default as SyncLoopHook } from './SyncLoopHook';

export { default as AsyncSeriesHook } from './AsyncSeriesHook';
export { default as AsyncSeriesBailHook } from './AsyncSeriesBailHook';
export { default as AsyncSeriesWaterfallHook } from './AsyncSeriesWaterfallHook';
export { default as AsyncParallelHook } from './AsyncParallelHook';
export { default as AsyncParallelBailHook } from './AsyncParallelBailHook';
