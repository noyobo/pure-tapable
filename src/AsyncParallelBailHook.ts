import { Async } from './Hook.js';
import type { DoneCallback, TapAsyncFn, TapPromiseFn } from './types';

/**
 * 并行串行 hook
 * 任一函数的返回非空值，则跳过剩余函数
 *
 * FIXME: 实现并行可中断
 */
export default class AsyncParallelBailHook<A extends unknown[]> extends Async<A, TapAsyncFn<A>, TapPromiseFn<A>> {
  callAsync = (...args: [...A, DoneCallback]) => {
    let taskIndex = 0;
    const asyncTaps = this.asyncTaps;
    const taskLength = asyncTaps.length;

    const done = args.pop() as DoneCallback;
    const params = this.argsToParams(args as unknown as A);

    const next = (error?: any, result?: any) => {
      if (error != undefined) return done(error);
      if (typeof result !== 'undefined') return done(null, result);

      if (taskIndex === taskLength) {
        done();
      } else {
        const { fn } = asyncTaps[taskIndex];
        fn.apply(this, [
          ...params,
          (error, result) => {
            taskIndex += 1;
            next(error, result);
          },
        ]);
      }
    };
    next();
  };

  promise = (...args: A): Promise<any> => {
    const params = this.argsToParams(args);
    let taskIndex = 0;

    return new Promise((resolve, reject) => {
      const next = async (value?: any) => {
        if (taskIndex === this.promiseTaps.length) {
          return resolve(value);
        }
        const { fn } = this.promiseTaps[taskIndex];
        return await fn(...params);
      };

      const checkDone = (result?: any) => {
        if (typeof result !== 'undefined') {
          return resolve(result);
        }
        next()
          .then((value) => {
            taskIndex += 1;
            checkDone(value);
          })
          .catch((error) => {
            reject(error);
          });
      };

      checkDone();
    });
  };
}
