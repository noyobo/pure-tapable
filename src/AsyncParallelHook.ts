import { Async } from './Hook.js';
import type { DoneCallback, TapAsyncFn, TapPromiseFn } from './types';

/**
 * 异步并行 hook
 * 忽略钩子函数的返回值
 */
export default class AsyncParallelHook<A extends unknown[]> extends Async<A, TapAsyncFn<A>, TapPromiseFn<A>> {
  callAsync = (...args: [...A, DoneCallback]) => {
    const done = args.pop() as DoneCallback;
    const argv = args as unknown as A;
    let isComplete = false;
    let taskDone = this.asyncTaps.length;

    // 所有异步函数执行完成触发 callAsync 回调
    const checkDone = (error: any) => {
      if (!isComplete) {
        taskDone -= 1;
        if (error != undefined) {
          isComplete = true;
          done(error);
        } else if (taskDone == 0) {
          isComplete = true;
          done();
        }
      }
    };

    for (let i = 0; i < this.asyncTaps.length; i++) {
      const { fn } = this.asyncTaps[i];
      // @ts-ignore
      fn.apply(this, [
        ...argv,
        function taskCallback(error) {
          checkDone(error);
        },
      ]);
    }
  };

  promise = (...args: A): Promise<void[]> => {
    return Promise.all(
      this.promiseTaps.map(({ fn }) => {
        return fn.apply(this, args);
      }),
    );
  };
}
