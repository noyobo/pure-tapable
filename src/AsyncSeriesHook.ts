import { Async } from './Hook.js';
import { DoneCallback, TapAsyncFn, TapPromiseFn } from './types';

/**
 * 异步串行 hook
 * 忽略钩子函数的返回值
 */
export default class AsyncSeriesHook<A extends unknown[]> extends Async<A, TapAsyncFn<A>, TapPromiseFn<A>> {
  callAsync = (...args: [...A, DoneCallback]) => {
    let taskIndex = 0;
    const asyncTaps = this.asyncTaps;
    const taskLength = asyncTaps.length;

    const done = args.pop() as DoneCallback;
    const params = this.argsToParams(args as unknown as A);

    const next = (error?: any, value?: any) => {
      if (error != undefined) return done(error);

      if (taskIndex === taskLength) {
        done();
      } else {
        const { fn } = asyncTaps[taskIndex];
        fn.apply(this, [
          ...params,
          (error, result) => {
            taskIndex += 1;
            next(error, result);
          },
        ]);
      }
    };
    next();
  };

  promise = (...args: A): Promise<any> => {
    const params = this.argsToParams(args);
    return new Promise((resolve, reject) => {
      let taskIndex = 0;
      const next = async (value?: any) => {
        if (taskIndex === this.promiseTaps.length) {
          return resolve(value);
        }

        const { fn } = this.promiseTaps[taskIndex];
        return await fn.apply(this, params);
      };

      const checkDone = () => {
        next()
          .then((value) => {
            taskIndex += 1;
            checkDone();
          })
          .catch((error) => {
            reject(error);
          });
      };

      checkDone();
    });
  };
}
