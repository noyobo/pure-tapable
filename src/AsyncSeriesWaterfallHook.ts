import { Async } from './Hook.js';
import { DoneCallback, TapAsyncFn, TapPromiseFn } from './types.js';

/**
 * 异步串行 hook
 * 钩子函数的返回值作为下一个监听函数的第一个参数
 */
export default class AsyncSeriesWaterfallHook<A extends unknown[]> extends Async<A, TapAsyncFn<A>, TapPromiseFn<A>> {
  validateArgs() {
    if (this.args.length < 1) {
      throw new Error(`Waterfall hooks must have at least one argument`);
    }
  }

  callAsync = (...args: [...A, DoneCallback]) => {
    const done = args.pop() as DoneCallback;
    const params = this.argsToParams(args as unknown as A);
    const [first, ...otherParams] = params;

    let taskIndex = 0;
    let lastParam = first;
    let lastError;

    const checkDone = (error: any, result: any) => {
      taskIndex += 1;
      lastError = error;

      if (result !== undefined) {
        lastParam = result;
      }

      if (lastError instanceof Error) {
        done(lastError);
      } else if (taskIndex === this.asyncTaps.length) {
        done(null, lastParam);
      } else {
        next([lastParam, ...otherParams] as A);
      }
    };

    const next = (params: A) => {
      const { fn } = this.asyncTaps[taskIndex];
      fn.apply(this, [
        ...params,
        (error, result) => {
          checkDone(error, result);
        },
      ]);
    };
    next(params);
  };

  promise(...args: A): Promise<any> {
    const params = this.argsToParams(args);
    let taskIndex = 0;

    const [first, ...otherParams] = params;
    let lastParam = first;

    return new Promise((resolve, reject) => {
      const checkDone = (result: any) => {
        if (result !== undefined) {
          lastParam = result;
        }

        if (taskIndex === this.promiseTaps.length) {
          resolve(lastParam);
        } else {
          run([lastParam, ...otherParams] as A);
        }
      };

      const next = (args: A) => {
        const { fn } = this.promiseTaps[taskIndex];
        return fn.apply(this, args);
      };

      const run = (args: any) => {
        next(args)
          .then((result) => {
            taskIndex += 1;
            checkDone(result);
          })
          .catch(reject);
      };
      run([lastParam, ...otherParams] as A);
    });
  }
}
