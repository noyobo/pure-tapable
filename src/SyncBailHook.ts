type Result = void | unknown;
type TapFn<P extends unknown[]> = (...args: P) => Result;

import { Sync } from './Hook.js';

/**
 * 同步串行 hook
 * 任一函数的返回值不为 undefined，则跳过剩余函数
 */
export default class SyncBailHook<P extends unknown[]> extends Sync<P, TapFn<P>> {
  call(...args: P): Result {
    let result: Result;
    for (let i = 0; i < this.taps.length; i++) {
      const { fn } = this.taps[i];
      result = fn.apply(this, args);
      if (result != undefined) {
        break;
      }
    }
    return result;
  }
}
