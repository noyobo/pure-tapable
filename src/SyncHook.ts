import { Sync } from './Hook.js';

type TapFn<P extends unknown[]> = (...args: P) => void;

/**
 * 同步串行 hook
 * 忽略钩子函数的返回值
 */

export default class SyncHook<P extends unknown[]> extends Sync<P, TapFn<P>> {
  call(...args: P): void {
    const params = this.argsToParams(args);
    this.taps.forEach(({ fn }) => {
      fn.apply(this, params);
    });
  }
}
