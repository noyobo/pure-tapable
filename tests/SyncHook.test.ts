import { describe, expect, it } from 'vitest';
import { SyncHook as S } from 'tapable';

import { SyncHook } from '../src';

describe('SyncHook', function () {
  it('call multi', function () {
    function run(Hook: typeof SyncHook, log: any[] = []) {
      const queue = new Hook<[string, number]>(['name', 'value']);
      queue.tap('1', (name, value) => {
        log.push([name, value, 1]);
      });
      queue.tap('2', (name, value) => {
        log.push([name, value, 2]);
      });
      queue.tap('3', (name, value) => {
        log.push([name, value, 3]);
      });
      return [queue.call('hooks', 0), log];
    }
    const a = run(SyncHook);
    // @ts-ignore
    const b = run(S);
    expect(a[0]).toMatchSnapshot();
    expect(b[0]).toMatchSnapshot();
    expect(a[0]).toEqual(b[0]);
    expect(a[1]).toMatchSnapshot();
    expect(b[1]).toMatchSnapshot();
    expect(a[1]).toEqual(b[1]);
  });

  it('call single', function () {
    function run(Hook: typeof SyncHook, log: any[] = []) {
      const queue = new Hook<[string]>(['name']);
      queue.tap('1', (name) => {
        log.push([name, 1]);
      });
      queue.tap('2', (name) => {
        log.push([name, 2]);
      });
      queue.tap('3', (name) => {
        log.push([name, 3]);
      });
      return [queue.call('hooks'), log];
    }
    const a = run(SyncHook);
    // @ts-ignore
    const b = run(S);
    expect(a[0]).toMatchSnapshot();
    expect(b[0]).toMatchSnapshot();
    expect(a[0]).toEqual(b[0]);
    expect(a[1]).toMatchSnapshot();
    expect(b[1]).toMatchSnapshot();
    expect(a[1]).toEqual(b[1]);
  });

  it('call empty', function () {
    function run(Hook: typeof SyncHook, log: any[] = []) {
      const queue = new Hook<[]>();
      queue.tap('1', () => {
        log.push([1]);
      });
      queue.tap('2', () => {
        log.push([2]);
      });
      queue.tap('3', () => {
        log.push([3]);
      });
      return [queue.call(), log];
    }
    const a = run(SyncHook);
    // @ts-ignore
    const b = run(S);
    expect(a[0]).toMatchSnapshot();
    expect(b[0]).toMatchSnapshot();
    expect(a[0]).toEqual(b[0]);
    expect(a[1]).toMatchSnapshot();
    expect(b[1]).toMatchSnapshot();
    expect(a[1]).toEqual(b[1]);
  });
});
