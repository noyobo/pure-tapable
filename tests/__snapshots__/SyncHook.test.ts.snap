// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SyncHook call empty 1`] = `undefined`;

exports[`SyncHook call empty 2`] = `undefined`;

exports[`SyncHook call empty 3`] = `
Array [
  Array [
    1,
  ],
  Array [
    2,
  ],
  Array [
    3,
  ],
]
`;

exports[`SyncHook call empty 4`] = `
Array [
  Array [
    1,
  ],
  Array [
    2,
  ],
  Array [
    3,
  ],
]
`;

exports[`SyncHook call multi 1`] = `undefined`;

exports[`SyncHook call multi 2`] = `undefined`;

exports[`SyncHook call multi 3`] = `
Array [
  Array [
    "hooks",
    0,
    1,
  ],
  Array [
    "hooks",
    0,
    2,
  ],
  Array [
    "hooks",
    0,
    3,
  ],
]
`;

exports[`SyncHook call multi 4`] = `
Array [
  Array [
    "hooks",
    0,
    1,
  ],
  Array [
    "hooks",
    0,
    2,
  ],
  Array [
    "hooks",
    0,
    3,
  ],
]
`;

exports[`SyncHook call single 1`] = `undefined`;

exports[`SyncHook call single 2`] = `undefined`;

exports[`SyncHook call single 3`] = `
Array [
  Array [
    "hooks",
    1,
  ],
  Array [
    "hooks",
    2,
  ],
  Array [
    "hooks",
    3,
  ],
]
`;

exports[`SyncHook call single 4`] = `
Array [
  Array [
    "hooks",
    1,
  ],
  Array [
    "hooks",
    2,
  ],
  Array [
    "hooks",
    3,
  ],
]
`;
